README.md
## Mask-RCNN in Tensorflow v2  ##

This repository is based on [matterport](https://github.com/matterport/Mask_RCNN) Mask-RCNN model implementation. The
main things about the model were added from the original repository. The repo is an attempt to make Mask-RCNN model more
transparent to researchers and more applicable in terms of inference optimization. Besides, new backbones were added in
order to have a choice in balance between accuracy and speed, to make model more task-specific.

### Supported Tensorflow versions

* v2.2.0, v2.3.4, v2.4.3, v2.5.1

### Supported backbones ###

* ResNet [18, 34, 50, 101, 152]
* ResNeXt [50, 101]
* SE-ResNet [18, 34, 50, 101, 152]
* SE-ResNeXt [50, 101]
* SE-Net [154]
* MobileNet [V1, V2]
* EfficientNet [B0, B1, B2, B3, B4, B5, B6, B7]


      Backbone keys:

      'resnet18', 'resnet34', 'resnet50', 'resnet101', 'resnet152',
      'resnext50', 'resnext101',
      'seresnet18', 'seresnet34', 'seresnet50', 'seresnet101', 'seresnet152',
      'seresnext50', 'seresnext101', 'senet154',
      'mobilenet', 'mobilenetv2',
      'efficientnetb0', 'efficientnetb1', 'efficientnetb2', 'efficientnetb3',
      'efficientnetb4', 'efficientnetb5', 'efficientnetb6', 'efficientnetb7',


## Getting started

### Environment setup

```bash
# Define preferred Tensorflow version: tf2.{2,3,4,5}
# For example, Tensorflow 2.2 env:
$ conda create -n tf2.2 python=3.7
$ conda activate tf2.2
$ pip install -r requirements_tf2.2.txt

# You may also need onnx_graphsurgeon and tensorrt python binding for TensorRT optimization
$ pip install <TENSORRT_PATH>/python/<TENSORRT_PYTHON_BINDING.whl>
$ pip install <TENSORRT_PATH>/onnx_graphsurgeon/onnx_graphsurgeon-x.y.z-py2.py3-none-any.whl
```

### Prepare dataset class and data augmentation

1. There is a general config about Mask-RCNN building and training in `./src/common/config.py` represented as a dict.
   Prepare it for a specific task (`CLASS_DICT` dictionary for class ids and names, other parameters are in `CONFIG`
   dictionary.)

2. Configure your dataset class. In the basic example we use general dataset class named `SegmentationDataset` for
   dealing with masks made in VGG Image Annotator.\
   In `./src/samples/balloon` you can inspect prepared `BalloonDataset` which inherits `SegmentationDataset` and process
   balloon image samples from the original repository.\
   In `./src/samples/coco` you can inspect prepared `CocoDataset` for MS COCO dataset.\
   Any prepared dataset class can be passed to DataLoader in  `./src/preprocess/preprocess.py` which generates batches.\
   You can also configure your own data augmentation. The default training augmentation is
   in `./src/preprocess/augmentation.py`
   in `get_training_augmentation` function. The default augmentation pipeline is based
   on [albumentations](https://github.com/albumentations-team/albumentations) library.

   See also:
    * `./notebooks/example_data_loader_balloon.ipynb`
    * `./notebooks/example_data_loader_coco.ipynb`

### Training

Basic example:

```python
import tensorflow as tf
from preprocess import preprocess
from preprocess import augmentation as aug
from training import train_model
from model import mask_rcnn_functional
from common.utils import tf_limit_gpu_memory
from common.config import CONFIG

# Limit GPU memory for tensorflow container
tf_limit_gpu_memory(tf, 4500)

# Update info about classes in your dataset
CONFIG.update({'class_dict': {},
               'num_classes':,
},
)
CONFIG.update({'meta_shape': (1 + 3 + 3 + 4 + 1 + CONFIG['num_classes']), })

# Init Mask-RCNN model
model = mask_rcnn_functional(config=CONFIG)

# Init training and validation datasets
base_dir = os.getcwd().replace('src', 'dataset_folder')
train_dir = os.path.join(base_dir, 'train')
val_dir = os.path.join(base_dir, 'val')

train_dataset = preprocess.SegmentationDataset(images_dir=train_dir,
                                               classes_dict=CONFIG['class_dict'],
                                               preprocess_transform=preprocess.get_input_preprocess(
                                                   normalize=CONFIG['normalization']
                                               ),
                                               augmentation=aug.get_training_augmentation(),
                                               **CONFIG
                                                )
val_dataset = preprocess.SegmentationDataset(images_dir=val_dir,
                                             classes_dict=CONFIG['class_dict'],
                                             preprocess_transform=preprocess.get_input_preprocess(
                                                 normalize=CONFIG['normalization']
                                             ),
                                             json_annotation_key=None,
                                             **CONFIG
                                             )
# train_model function includes dataset and dataloader initialization, callbacks configuration, 
# a list of losses definition and final model compiling with optimizer defined in CONFIG.
train_model(model,
            train_dataset=train_dataset,
            val_dataset=val_dataset,
            config=CONFIG,
            weights_path=None)
```

Balloon dataset example:

Download balloon dataset [here](https://github.com/matterport/Mask_RCNN/releases/download/v2.1/balloon_dataset.zip)

```python
import os

os.chdir('..')
import tensorflow as tf

from samples.balloon import balloon
from preprocess import preprocess
from preprocess import augmentation as aug
from training import train_model
from model import mask_rcnn_functional
from common.utils import tf_limit_gpu_memory

# Limit GPU memory for tensorflow container
tf_limit_gpu_memory(tf, 4500)

from common.config import CONFIG

CONFIG.update(balloon.BALLON_CONFIG)

# Init Mask-RCNN model
model = mask_rcnn_functional(config=CONFIG)

# Init training and validation datasets
base_dir = os.getcwd().replace('src', 'balloon')
train_dir = os.path.join(base_dir, 'train')
val_dir = os.path.join(base_dir, 'val')

train_dataset = balloon.BalloonDataset(images_dir=train_dir,
                                       class_key='object',
                                       classes_dict=CONFIG['class_dict'],
                                       preprocess_transform=preprocess.get_input_preprocess(
                                           normalize=CONFIG['normalization']
                                       ),
                                       augmentation=aug.get_training_augmentation(),
                                       json_annotation_key=None,
                                       **CONFIG
                                       )

val_dataset = balloon.BalloonDataset(images_dir=val_dir,
                                     class_key='object',
                                     classes_dict=CONFIG['class_dict'],
                                     preprocess_transform=preprocess.get_input_preprocess(
                                         normalize=CONFIG['normalization']
                                     ),
                                     json_annotation_key=None,
                                     **CONFIG
                                     )

# train_model function includes dataset and dataloader initialization, callbacks configuration, 
# a list of losses definition and final model compiling with optimizer defined in CONFIG.
train_model(model,
            train_dataset=train_dataset,
            val_dataset=val_dataset,
            config=CONFIG,
            weights_path=None)
```

See `./notebooks/example_training_balloon.ipynb`.

MS COCO dataset example:

```python
import os

os.chdir('..')
import tensorflow as tf

from samples.coco import coco
from preprocess import preprocess
from preprocess import augmentation as aug
from training import train_model
from model import mask_rcnn_functional
from common.utils import tf_limit_gpu_memory

# Limit GPU memory for tensorflow container
tf_limit_gpu_memory(tf, 4500)

from common.config import CONFIG

CONFIG.update(coco.COCO_CONFIG)

# Init Mask-RCNN model
model = mask_rcnn_functional(config=CONFIG)

# You can also download dataset with auto_download=True argument
# It will be downloaded and unzipped in dataset_dir
base_dir = r'<COCO_PATH>/coco2017'
train_dir = os.path.join(base_dir, 'train')
val_dir = os.path.join(base_dir, 'val')

train_dataset = coco.CocoDataset(dataset_dir=base_dir,
                                 subset='train',
                                 year=2017,
                                 auto_download=True,
                                 preprocess_transform=preprocess.get_input_preprocess(
                                     normalize=CONFIG['normalization']
                                 ),
                                 augmentation=aug.get_training_augmentation(),
                                 **CONFIG
                                 )

val_dataset = coco.CocoDataset(dataset_dir=base_dir,
                               subset='val',
                               year=2017,
                               auto_download=True,
                               preprocess_transform=preprocess.get_input_preprocess(
                                   normalize=CONFIG['normalization']
                               ),
                               **CONFIG
                               )

train_model(model,
            train_dataset=train_dataset,
            val_dataset=val_dataset,
            config=CONFIG,
            weights_path=None)
```

See `./notebooks/example_training_coco.ipynb`.

4. Logs folder with weights and scalars will appear outside `src`. Monitoring training with tensorboard tool:

```bash
$ tensorboard --log_dir=logs
```

### Inference

See inference example in `./notebooks/example_inference_tf_onnx_trt_balloon.ipynb`

### Inference optimization

The project suggests a straightforward way of __Mask-RCNN inference optimization__ on x86_64 architecture and also on
NVIDIA Jetson devices (AArch64). Here you do not need to fix .uff graph and then optimize it with TensorRT. The model
optimizing way here is based on pure .onnx graph with only one prepared .onnx graph modification function for TensorRT.
You can inspect optimization steps with python in `example_tensorflow_to_onnx_tensorrt_balloon.ipynb`.

Optimization steps:

1. Initialize your model in inference mode and load its weights. Thus, your model won't include unnecessary layers that
   is used in training mode.
2. Convert your tensorflow.keras model to .onnx with `tf2onnx`.

__Inference with onnxruntime:__  
From this step, you can use generated .onnx graph in `onnxruntime` and `onnxruntime-gpu` inference.

__Inference with TensorRT:__

3. Change your .onnx graph made on step 2. by including TensorRT-implemented Mask-RCNN layers with

CHANGELOG.md
# Change Log
All notable changes to the maskrcnn with tensorflow 2 will be documented in this file.

 
## [Unreleased] - 2021-09-25

### Added
- ResNet's backbones got `resnet_leaky_relu` option.
- Mask and classifier heads also got `mask_head_leaky_relu` and `cls_head_leaky_relu` options.
- Extended ResNets and EfficientNets backbones: added ResNet152, EfficientNet [B4, B5, B6, B7] backbones.
- ResNeXt backbones [50, 101].
- SE-ResNet backbones [18, 34, 50, 101, 152].
- SE-ResNeXt backbones [50, 101].
- SE-Net backbones [154].
 
### Changed
- New tensorboard log folder format: `[maskrcnn_<BACKBONE_NAME>_<YYYY-MM-DD_hh-mm-ss>]`
- `logs` folder now is generated outside `src`
- Updated `README.md`
 
### Fixed

* Normalization issue. Now normalization can be specified via model config in `src/common/config.py`. 
* ONNX ZeroPadding for TensorRT engines.

### Known issues

- There is a drop in performance after ONNX modification for TRT, also NaNs happens in TensorRT model output.
  The issue is under research.


- TensorRT: `../rtSafe/cublas/cublasLtWrapper.cpp (279) - Assertion Error in getCublasLtHeuristic: 0 (cublasStatus == CUBLAS_STATUS_SUCCESS)`
   
    * Possible cause: cuda and TensorRT versions mismatch;
    * Possible workaround if error still exists - remove cublasLt from tacticSources:
       * fp32:
          `trtexec --onnx=<PATH_TO_ONNX_GRAPH> --saveEngine=<PATH_TO_TRT_ENGINE> --tacticSources=-cublasLt,+cublas --workspace=<WORKSPACE_SIZE> --verbose`
       * fp16:
           `trtexec --onnx=<PATH_TO_ONNX_GRAPH> --saveEngine=<PATH_TO_TRT_ENGINE> --tacticSources=-cublasLt,+cublas --fp16 --workspace=<WORKSPACE_SIZE> --verbose`
    

- Tensorlfow v2.5: AttributeError: module 'keras.utils' has no attribute 'get_file'
       
   * Possible cause: the influence of keras-nightly automatically suggested with tensorflow installation.
   * Possible workaround:
         
     Open `__init__.py` in `<ANACONDA_PATH>/envs/tf2.5/lib/python3.8/site-packages/classification_models`
         
     Modify file:
     ```python
     import keras_applications as ka
     from .__version__ import __version__
     import tensorflow.keras.utils as utils
         
     def get_submodules_from_kwargs(kwargs):
         backend = kwargs.get('backend', ka._KERAS_BACKEND)
         layers = kwargs.get('layers', ka._KERAS_LAYERS)
         models = kwargs.get('models', ka._KERAS_MODELS)
         return backend, layers, models, utils
     ```
   
- `senet154`, `efficientnetb5`, `efficientnetb6`, `efficientnetb7` backbones are not tested enough for now because of the 
  high GPU memory consumption. Thus, the model with `senet154` backbone may be not supported for ONNX graph modification.

## [Unreleased] - 2021-12-17

### Fixed

* `steps_per_epoch` calculation in  `src/preprocess/preprocess.py`. 
   [Related issue](https://github.com/alexander-pv/maskrcnn_tf2/issues/9).

src/samples/coco/coco.py
"""
Mask R-CNN
Configurations and data loading code for MS COCO.
Copyright (c) 2017 Matterport, Inc.
Licensed under the MIT License (see LICENSE for details)
Written by Waleed Abdulla
"""

import os
import shutil
import time
import urllib.request
import zipfile

import numpy as np
import skimage
import skimage.io
import tqdm
from preprocess import preprocess
from pycocotools import mask as maskUtils
# Download and install the Python COCO tools from https://github.com/waleedka/coco
# That's a fork from the original https://github.com/pdollar/coco with a bug
# fix for Python 3.
# I submitted a pull request https://github.com/cocodataset/cocoapi/pull/50
# If the PR is merged then use the original repo.
# Note: Edit PythonAPI/Makefile and replace "python" with "python3".
from pycocotools.coco import COCO
from pycocotools.cocoeval import COCOeval

COCO_CONFIG = {'class_dict': {'background': 0,
                              'person': 1,
                              'bicycle': 2,
                              'car': 3,
                              'motorcycle': 4,
                              'airplane': 5,
                              'bus': 6,
                              'train': 7,
                              'truck': 8,
                              'boat': 9,
                              'traffic light': 10,
                              'fire hydrant': 11,
                              'stop sign': 12,
                              'parking meter': 13,
                              'bench': 14,
                              'bird': 15,
                              'cat': 16,
                              'dog': 17,
                              'horse': 18,
                              'sheep': 19,
                              'cow': 20,
                              'elephant': 21,
                              'bear': 22,
                              'zebra': 23,
                              'giraffe': 24,
                              'backpack': 25,
                              'umbrella': 26,
                              'handbag': 27,
                              'tie': 28,
                              'suitcase': 29,
                              'frisbee': 30,
                              'skis': 31,
                              'snowboard': 32,
                              'sports ball': 33,
                              'kite': 34,
                              'baseball bat': 35,
                              'baseball glove': 36,
                              'skateboard': 37,
                              'surfboard': 38,
                              'tennis racket': 39,
                              'bottle': 40,
                              'wine glass': 41,
                              'cup': 42,
                              'fork': 43,
                              'knife': 44,
                              'spoon': 45,
                              'bowl': 46,
                              'banana': 47,
                              'apple': 48,
                              'sandwich': 49,
                              'orange': 50,
                              'broccoli': 51,
                              'carrot': 52,
                              'hot dog': 53,
                              'pizza': 54,
                              'donut': 55,
                              'cake': 56,
                              'chair': 57,
                              'couch': 58,
                              'potted plant': 59,
                              'bed': 60,
                              'dining table': 61,
                              'toilet': 62,
                              'tv': 63,
                              'laptop': 64,
                              'mouse': 65,
                              'remote': 66,
                              'keyboard': 67,
                              'cell phone': 68,
                              'microwave': 69,
                              'oven': 70,
                              'toaster': 71,
                              'sink': 72,
                              'refrigerator': 73,
                              'book': 74,
                              'clock': 75,
                              'vase': 76,
                              'scissors': 77,
                              'teddy bear': 78,
                              'hair drier': 79,
                              'toothbrush': 80,
                              },
               'num_classes': 80 + 1,
               }

COCO_CONFIG.update({'meta_shape': 1 + 3 + 3 + 4 + 1 + COCO_CONFIG['num_classes']})


class CocoDataset(preprocess.SegmentationDataset):

    def __init__(self, dataset_dir, subset, year, class_ids=None,
                 class_map=None, return_coco=False, auto_download=False, **kwargs):
        """
        CocoDataset class for coco data management.
        Args:
            dataset_dir: Directory where to search or download COCO dataset
            subset:      What to load (train, val, minival, valminusminival)
            year:        What dataset year to load (2014, 2017) as a string, not an integer
            class_ids: If provided, only loads images that have the given classes.
            class_map: TODO: Not implemented yet. Supports mapping classes from
                       different datasets to the same class ID.
            return_coco: If True, returns the COCO object.
            auto_download: Automatically download and unzip MS-COCO images and annotations
        """
        # super(preprocess.SegmentationDataset, self).__init__(**kwargs)
        super().__init__(**kwargs)
        self._image_ids = []
        self.image_info = []
        # Background is always the first class
        self.class_info = [{"source": "", "id": 0, "name": "BG"}]
        self.source_class_ids = {}
        self.subset = subset

        # Load and prepare dataset during initialization
        self.load_coco(dataset_dir=dataset_dir, subset=subset, year=year, class_ids=class_ids,
                       class_map=class_map, return_coco=return_coco, auto_download=auto_download)
        self.prepare()

    def add_class(self, source, class_id, class_name):
        assert "." not in source, "Source name cannot contain a dot"
        # Does the class exist already?
        for info in self.class_info:
            if info['source'] == source and info["id"] == class_id:
                # source.class_id combination already available, skip
                return
        # Add the class
        self.class_info.append({
            "source": source,
            "id": class_id,
            "name": class_name,
        })

    def add_image(self, source, image_id, path, **kwargs):
        image_info = {
            "id": image_id,
            "source": source,
            "path": path,
        }
        image_info.update(kwargs)
        self.image_info.append(image_info)

    def map_source_class_id(self, source_class_id):
        """Takes a source class ID and returns the int class ID assigned to it.
        For example:
        dataset.map_source_class_id("coco.12") -> 23
        """
        return self.class_from_source_map[source_class_id]

    @property
    def image_ids(self):
        return self._image_ids

    def prepare(self, class_map=None):
        """Prepares the Dataset class for use.
        TODO: class map is not supported yet. When done, it should handle mapping
              classes from different datasets to the same class ID.
        """

        def clean_name(name):
            """Returns a shorter version of object names for cleaner display."""
            return ",".join(name.split(",")[:1])

        # Build (or rebuild) everything else from the info dicts.
        self.num_classes = len(self.class_info)
        self.class_ids = np.arange(self.num_classes)
        self.class_names = [clean_name(c["name"]) for c in self.class_info]
        self.num_images = len(self.image_info)
        self._image_ids = np.arange(self.num_images)

        # Mapping from source class and image IDs to internal IDs
        self.class_from_source_map = {"{}.{}".format(info['source'], info['id']): id
                                      for info, id in zip(self.class_info, self.class_ids)}
        self.image_from_source_map = {"{}.{}".format(info['source'], info['id']): id
                                      for info, id in zip(self.image_info, self.image_ids)}

        # Map sources to class_ids they support
        self.sources = list(set([i['source'] for i in self.class_info]))
        self.source_class_ids = {}
        # Loop over datasets
        for source in self.sources:
            self.source_class_ids[source] = []
            # Find classes that belong to this dataset
            for i, info in enumerate(self.class_info):
                # Include BG class in all datasets
                if i == 0 or source == info['source']:
                    self.source_class_ids[source].append(i)
        self.images_names = self._image_ids  # Fix for DataLoader

    def load_coco(self, dataset_dir, subset, year, class_ids=None,
                  class_map=None, return_coco=False, auto_download=False):
        """Load a subset of the COCO dataset.
        dataset_dir: The root directory of the COCO dataset.
        subset: What to load (train, val, minival, valminusminival)
        year: What dataset year to load (2014, 2017) as a string, not an integer
        class_ids: If provided, only loads images that have the given classes.
        class_map: TODO: Not implemented yet. Supports mapping classes from
                   different datasets to the same class ID.
        return_coco: If True, returns the COCO object.
        auto_download: Automatically download and unzip MS-COCO images and annotations
        """

        if auto_download is True:
            self.auto_download(dataset_dir, subset, year)

        coco = COCO("{}/annotations/instances_{}{}.json".format(dataset_dir, subset, year))
        if subset == "minival" or subset == "valminusminival":
            subset = "val"
        image_dir = "{}/{}{}".format(dataset_dir, subset, year)

        # Load all classes or a subset?
        if not class_ids:
            # All classes
            class_ids = sorted(coco.getCatIds())

        # All images or a subset?
        if class_ids:
            image_ids = []
            for id in class_ids:
                image_ids.extend(list(coco.getImgIds(catIds=[id])))
            # Remove duplicates
            image_ids = list(set(image_ids))
        else:
            # All images
            image_ids = list(coco.imgs.keys())

        # Add classes
        for i in class_ids:
            self.add_class("coco", i, coco.loadCats(i)[0]["name"])

        # Add images
        for i in tqdm.tqdm(image_ids):
            self.add_image(
                "coco", image_id=i,
                path=os.path.join(image_dir, coco.imgs[i]['file_name']),
                width=coco.imgs[i]["width"],
                height=coco.imgs[i]["height"],
                annotations=coco.loadAnns(coco.getAnnIds(
                    imgIds=[i], catIds=class_ids, iscrowd=None)))
        if return_coco:
            return coco

    def auto_download(self, dataDir, dataType, dataYear):
        """Download the COCO dataset/annotations if requested.
        dataDir: The root directory of the COCO dataset.
        dataType: What to load (train, val, minival, valminusminival)
        dataYear: What dataset year to load (2014, 2017) as a string, not an integer
        Note:
            For 2014, use "train", "val", "minival", or "valminusminival"
            For 2017, only "train" and "val" annotations are available
        """

        # Setup paths and file names
        if dataType == "minival" or dataType == "valminusminival":
            imgDir = "{}/{}{}".format(dataDir, "val", dataYear)
            imgZipFile = "{}/{}{}.zip".format(dataDir, "val", dataYear)
            imgURL = "http://images.cocodataset.org/zips/{}{}.zip".format("val", dataYear)
        else:
            imgDir = "{}/{}{}".format(dataDir, dataType, dataYear)
            imgZipFile = "{}/{}{}.zip".format(dataDir, dataType, dataYear)
            imgURL = "http://images.cocodataset.org/zips/{}{}.zip".format(dataType, dataYear)
        # print("Image paths:"); print(imgDir); print(imgZipFile); print(imgURL)

        # Create main folder if it doesn't exist yet
        if not os.path.exists(dataDir):
            os.makedirs(dataDir)

        # Download images if not available locally
        if not os.path.exists(imgDir):
            os.makedirs(imgDir)
            print("Downloading images to " + imgZipFile + " ...")
            with urllib.request.urlopen(imgURL) as resp, open(imgZipFile, 'wb') as out:
                shutil.copyfileobj(resp, out)
            print("... done downloading.")
            print("Unzipping " + imgZipFile)
            with zipfile.ZipFile(imgZipFile, "r") as zip_ref:
                zip_ref.extractall(dataDir)
            print("... done unzipping")
        print("Will use images in " + imgDir)

        # Setup annotations data paths
        annDir = "{}/annotations".format(dataDir)
        if dataType == "minival":
            annZipFile = "{}/instances_minival2014.json.zip".format(dataDir)
            annFile = "{}/instances_minival2014.json".format(annDir)
            annURL = "https://dl.dropboxusercontent.com/s/o43o90bna78omob/instances_minival2014.json.zip?dl=0"
            unZipDir = annDir
        elif dataType == "valminusminival":
            annZipFile = "{}/instances_valminusminival2014.json.zip".format(dataDir)
            annFile = "{}/instances_valminusminival2014.json".format(annDir)
            annURL = "https://dl.dropboxusercontent.com/s/s3tw5zcg7395368/instances_valminusminival2014.json.zip?dl=0"
            unZipDir = annDir
        else:
            annZipFile = "{}/annotations_trainval{}.zip".format(dataDir, dataYear)
            annFile = "{}/instances_{}{}.json".format(annDir, dataType, dataYear)
            annURL = "http://images.cocodataset.org/annotations/annotations_trainval{}.zip".format(dataYear)
            unZipDir = dataDir
        # print("Annotations paths:"); print(annDir); print(annFile); print(annZipFile); print(annURL)

        # Download annotations if not available locally
        if not os.path.exists(annDir):
            os.makedirs(annDir)
        if not os.path.exists(annFile):
            if not os.path.exists(annZipFile):
                print("Downloading zipped annotations to " + annZipFile + " ...")
                with urllib.request.urlopen(annURL) as resp, open(annZipFile, 'wb') as out:
                    shutil.copyfileobj(resp, out)
                print("... done downloading.")
            print("Unzipping " + annZipFile)
            with zipfile.ZipFile(annZipFile, "r") as zip_ref:
                zip_ref.extractall(unZipDir)
            print("... done unzipping")
        print("Will use annotations in " + annFile)

    def load_image(self, image_id):
        """Load the specified image and return a [H,W,3] Numpy array.
        """
        # Load image
        image = skimage.io.imread(self.image_info[image_id]['path'])
        # If grayscale. Convert to RGB for consistency.
        if image.ndim != 3:
            image = skimage.color.gray2rgb(image)
        # If has an alpha channel, remove it for consistency
        if image.shape[-1] == 4:
            image = image[..., :3]
        return image

    def create_mask(self, image, image_id):
        """Load instance masks for the given image.
        Different datasets use different ways to store masks. This
        function converts the different mask format to one format
        in the form of a bitmap [height, width, instances].
        Returns:
        masks: A bool array of shape [height, width, instance count] with
            one mask per instance.
        class_ids: a 1D array of class IDs of the instance masks.
        """
        # If not a COCO image, delegate to parent class.
        image_info = self.image_info[image_id]
        if image_info["source"] != "coco":
            return super(CocoDataset, self).create_mask(image, image_id)

        instance_masks = []
        class_ids = []
        annotations = self.image_info[image_id]["annotations"]
        # Build mask of shape [height, width, instance_count] and list
        # of class IDs that correspond to each channel of the mask.
        for annotation in annotations:
            class_id = self.map_source_class_id(
                "coco.{}".format(annotation['category_id']))
            if class_id:
                m = self.annToMask(annotation, image_info["height"],
                                   image_info["width"])
                # Some objects are so small that they're less than 1 pixel area
                # and end up rounded out. Skip those objects.
                if m.max() < 1:
                    continue
                # Is it a crowd? If so, use a negative class ID.
                if annotation['iscrowd']:
                    # Use negative class ID for crowds
                    class_id *= -1
                    # For crowd masks, annToMask() sometimes returns a mask
                    # smaller than the given dimensions. If so, resize it.
                    if m.shape[0] != image_info["height"] or m.shape[1] != image_info["width"]:
                        m = np.ones([image_info["height"], image_info["width"]], dtype=bool)
                instance_masks.append(m)
                class_ids.append(class_id)

        # Pack instance masks into an array
        if class_ids:
            mask = np.stack(instance_masks, axis=2).astype(np.bool)
            class_ids = np.array(class_ids, dtype=np.int32)
            return mask, class_ids
        else:
            # Call super class to return an empty mask
            return super(CocoDataset, self).create_mask(image_id)

    def image_reference(self, image_id):
        """Return a link to the image in the COCO Website."""
        info = self.image_info[image_id]
        if info["source"] == "coco":
            return "http://cocodataset.org/#explore?id={}".format(info["id"])
        else:
            super(CocoDataset, self).image_reference(image_id)

    # The following two functions are from pycocotools with a few changes.

    def annToRLE(self, ann, height, width):
        """
        Convert annotation which can be polygons, uncompressed RLE to RLE.
        :return: binary mask (numpy 2D array)
        """
        segm = ann['segmentation']
        if isinstance(segm, list):
            # polygon -- a single object might consist of multiple parts
            # we merge all parts into one mask rle code
            rles = maskUtils.frPyObjects(segm, height, width)
            rle = maskUtils.merge(rles)
        elif isinstance(segm['counts'], list):
            # uncompressed RLE
            rle = maskUtils.frPyObjects(segm, height, width)
        else:
            # rle
            rle = ann['segmentation']
        return rle

    def annToMask(self, ann, height, width):
        """
        Convert annotation which can be polygons, uncompressed RLE, or RLE to binary mask.
        :return: binary mask (numpy 2D array)
        """
        rle = self.annToRLE(ann, height, width)
        m = maskUtils.decode(rle)
        return m


def build_coco_results(dataset, image_ids, rois, class_ids, scores, masks):
    """Arrange resutls to match COCO specs in http://cocodataset.org/#format
    """
    # If no results, return an empty list
    if rois is None:
        return []

    results = []
    for image_id in image_ids:
        # Loop through detections
        for i in range(rois.shape[0]):
            class_id = class_ids[i]
            score = scores[i]
            bbox = np.around(rois[i], 1)
            mask = masks[:, :, i]

            result = {
                "image_id": image_id,
                "category_id": dataset.get_source_class_id(class_id, "coco"),
                "bbox": [bbox[1], bbox[0], bbox[3] - bbox[1], bbox[2] - bbox[0]],
                "score": score,
                "segmentation": maskUtils.encode(np.asfortranarray(mask))
            }
            results.append(result)
    return results


def evaluate_coco(model, dataset, coco, eval_type="bbox", limit=0, image_ids=None):
    """Runs official COCO evaluation.
    dataset: A Dataset object with valiadtion data
    eval_type: "bbox" or "segm" for bounding box or segmentation evaluation
    limit: if not 0, it's the number of images to use for evaluation
    """
    # Pick COCO images from the dataset
    image_ids = image_ids or dataset.image_ids

    # Limit to a subset
    if limit:
        image_ids = image_ids[:limit]

    # Get corresponding COCO image IDs.
    coco_image_ids = [dataset.image_info[id]["id"] for id in image_ids]

    t_prediction = 0
    t_start = time.time()

    results = []
    for i, image_id in enumerate(image_ids):
        # Load image
        image = dataset.load_image(image_id)

        # Run detection
        t = time.time()
        r = model.detect([image], verbose=0)[0]
        t_prediction += (time.time() - t)

        # Convert results to COCO format
        # Cast masks to uint8 because COCO tools errors out on bool
        image_results = build_coco_results(dataset, coco_image_ids[i:i + 1],
                                           r["rois"], r["class_ids"],
                                           r["scores"],
                                           r["masks"].astype(np.uint8))
        results.extend(image_results)

    # Load results. This modifies results with additional attributes.
    coco_results = coco.loadRes(results)

    # Evaluate
    cocoEval = COCOeval(coco, coco_results, eval_type)
    cocoEval.params.imgIds = coco_image_ids
    cocoEval.evaluate()
    cocoEval.accumulate()
    cocoEval.summarize()

    print("Prediction time: {}. Average {}/image".format(
        t_prediction, t_prediction / len(image_ids)))
    print("Total time: ", time.time() - t_start)
